src/mbgl/storage/file_source_request.hpp

```diff
@@ -0,0 +1,31 @@
+#pragma once
+
+#include <mbgl/storage/file_source.hpp>
+#include <mbgl/util/async_request.hpp>
+
+#include <memory>
+#include <functional>
+
+namespace mbgl {
+
+class Mailbox;
+
+class FileSourceRequest : public AsyncRequest {
+public:
+    FileSourceRequest(std::shared_ptr<Mailbox> mailbox);
+    ~FileSourceRequest() final;
+
+    void onResponse(FileSource::Callback&& callback);
+    void onCancel(std::function<void()>&& callback);
+    void setResponse(const Response& res);
+
+private:
+    void runResponseCallback();
+
+    std::shared_ptr<Mailbox> mailbox;
```

Comment:
Let's do like you suggested. I have the `Mailbox` on the constructor because right now we assume the scheduler to be a `RunLoop` but that will have to change if we want to make requests from a worker in a `ThreadPool`. Better keep that to another PR.